### PR TITLE
Add solutions to Assignment2

### PR DESCRIPTION
**Pull request by elifguler96**

### CODE REVIEW

File: Assignment2Q2.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		if (leftCA != null) {
+			return leftCA;
+		}
+
+		// both are on the right subtree
+		return rightCA;
+	}
+
+	/**
+	 * Checks is node is present in the tree whose root is root
```
Comment (katermar) [inline Assignment2Q2.java]: maybe there is a typo? "if"?

File: Assignment2Q1.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public class Assignment2Q1 {
+	public static void main(String[] args) {
+		BinaryTreeNode<Integer> n1 = new BinaryTreeNode<>(1, null, null);
```
Comment (katermar) [inline Assignment2Q1.java]: Isn't it better to use the contructor to create a simple node without linking on the left/right?

**Review by katermar — COMMENTED** [review]
Pretty good, I've leaved some comments, all about design

File: Assignment2Q1.java (hunk 2 above)
Comment (elifguler96) [reply Assignment2Q1.java]: But I use those links to create the tree. If I don't link the nodes, I won't have a binary tree.

File: Assignment2Q2.java (hunk 1 above)
Comment (elifguler96) [reply Assignment2Q2.java]: Yeah you are right 😅

File: Assignment2Q1.java (hunk 2 above)
Comment (katermar) [reply Assignment2Q1.java]: I mean, that you can avoid writing _null_'s there. Cuz with your 1-arg. costructor it may be the same.

Comment (elifguler96) [reply Assignment2Q1.java]: Oh yes you are right, but I just wanted all to be in the same format that's all :D

Comment (katermar) [reply Assignment2Q1.java]: ok, so in this case you just don't need that 1-arg. constructor :)

Comment (elifguler96) [reply Assignment2Q1.java]: Yeah I don't, just wrote it out of habit :D

File: Assignment2Q2.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		BinaryTreeNode<Integer> n2 = new BinaryTreeNode<>(5, null, null);
+		BinaryTreeNode<Integer> n3 = new BinaryTreeNode<>(3, n1, n2);
+		BinaryTreeNode<Integer> n4 = new BinaryTreeNode<>(14, null, null);
+		BinaryTreeNode<Integer> n5 = new BinaryTreeNode<>(9, n3, n4);
+		BinaryTreeNode<Integer> n6 = new BinaryTreeNode<>(19, null, null);
+		BinaryTreeNode<Integer> n7 = new BinaryTreeNode<>(18, null, n6);
+		BinaryTreeNode<Integer> n8 = new BinaryTreeNode<>(16, n5, n7);
+		BinaryTreeNode<Integer> n9 = new BinaryTreeNode<>(186, null, null);
+
+		System.out.println(findCommonAncestors(n9, n1, n8).getData());
```
Comment (brzozan) [inline Assignment2Q2.java]: While relying on printed output that has to be verified by human is very useful during development, it typically covers only one scenario and does not scale well.

The code for findCommonAncestors has several edge cases (value not present in the tree, one of the keys being ancestor of the other etc), so it would be a good practice to cover more of them with unit tests. Would you mind adding them, so that when the code is run it gives information about status of different test cases?

You could use JUnit for it or go with simple test method(s) that would execute findCommonAncestors for different scenarios (possibly even using the same tree) and report whether result was as expected or not?

File: Assignment2Q2.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+		if (node == null) {
+			throw new IllegalArgumentException();
+		}
+
+		// if node1 is the common ancestor, check if node2 is in the tree which
+		// has root node1. Also handles the case if node1 and node2 are the
+		// same.
+		if (node.getData().equals(node1.getData())) {
+			if (node1.getData().equals(node2.getData()) || contains(node2, node)) {
```
Comment (brzozan) [inline Assignment2Q2.java]: The node.getData().equals(node2.getData()) pattern appears quite a lot across the code. You could consider adding something like containsData(T data) method to node and use it instead.

File: Assignment2Q2.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		if (leftCA != null && rightCA != null) {
+			return root;
+		}
+
+		// both are on the left sub tree
+		if (leftCA != null) {
+			return leftCA;
+		}
+
+		// both are on the right subtree
```
Comment (brzozan) [inline Assignment2Q2.java]: ... or key does not exist in the tree (rightCA == null).

File: Assignment2Q2.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	 *            first node
+	 * @param node2
+	 *            second node
+	 * @param root
+	 *            root of the tree
+	 * @return lowest common ancestor of node1 and node2
+	 * @throws IllegalArgumentException
+	 *             if one or both of the nodes are not present in the tree
+	 */
+	public static <T> BinaryTreeNode<T> findCommonAncestors(BinaryTreeNode<T> node1, BinaryTreeNode<T> node2,
```
Comment (brzozan) [inline Assignment2Q2.java]: Optional thing to think about: could finding a common ancestor of two nodes be somehow simplified by reusing some code created to print ancestors of a single node?

File: Assignment2Q1.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	/**
+	 * Helper method of printAncestors
+	 * 
+	 * @param key
+	 *            key of the node whose ancestors are going to be printed
+	 * @param root
+	 *            root of the binary tree
+	 * @return list of ancestors of the node
+	 */
+	private static <T> List<T> findAncestors(T key, BinaryTreeNode<T> root) {
```
Comment (brzozan) [inline Assignment2Q1.java]: Small design comment: both findAncestors and findCommonAncestors operate on tree nodes and access their internal data (left, right etc). It is generally better to keep data and methods that operate on them together, so that it is clear that they depend on each other and to keep possible future changes in one place.

Would it make sense to move such functions and change them to (non-static) methods of BinaryTreeNode?

File: Assignment2Q1.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+			} else {
+				return null;
+			}
+		}
+
+		return ancestorList;
+	}
+}
+
+class BinaryTreeNode<T> {
```
Comment (brzozan) [inline Assignment2Q1.java]: BinaryTreeNode class is used by two other classes. It would be nice to put it in a separate .java file to make it more clear that it is shared rather than used just in Assignment2Q1.

File: Assignment2Q2.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	/**
+	 * Checks if node is present in the tree whose root is root
+	 * 
+	 * @param node
+	 *            node to be checked
+	 * @param root
+	 *            root of the tree
+	 * @return true if present, false otherwise
+	 */
+	private static <T> boolean contains(BinaryTreeNode<T> node, BinaryTreeNode<T> root) {
```
Comment (brzozan) [inline Assignment2Q2.java]: This also looks like something that would better fit into BinaryTreeNode class.

**Review by brzozan — CHANGES_REQUESTED** [review]
Very nice code.

I have commented on a few things that in my opinion could be improved from design and testing point of view.

File: Assignment2Q1.java (hunk 7 above)
Comment (elifguler96) [reply Assignment2Q1.java]: However, if I put them under BinaryTreeNode, I will have to have one .java file for both questions, I will not be able to separate them as Assignment2Q1 and Assignment2Q2. I thought the latter was a better convention.

File: Assignment2Q2.java (hunk 6 above)
Comment (elifguler96) [reply Assignment2Q2.java]: I first thought of that, but to use the code for findAncestors, I will have to store them in a list, which we are not allowed to do. That's why I did it like this.

File: Assignment2Q2.java (hunk 5 above)
Comment (elifguler96) [reply Assignment2Q2.java]: Is just changing the comment to "both are on the right subtree or key does not exist in the tree" ok or should I change the code too, to make it more readable?

File: Assignment2Q2.java (hunk 6 above)
Comment (brzozan) [reply Assignment2Q2.java]: Yup. I missed that requirement. Sorry about that.

File: Assignment2Q2.java (hunk 5 above)
Comment (brzozan) [reply Assignment2Q2.java]: Code is fine. Only the comment didn't match the actual situation.

**Comment by elifguler96** [comment]
I realized I forgot to add the BinaryTreeNode class 😅 I added it now.

File: Assignment2Q1Test.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        n2 = new BinaryTreeNode<>(5, null, null);
+        n3 = new BinaryTreeNode<>(3, n1, n2);
+        n4 = new BinaryTreeNode<>(14, null, null);
+        n5 = new BinaryTreeNode<>(9, n3, n4);
+        n6 = new BinaryTreeNode<>(19, null, null);
+        n7 = new BinaryTreeNode<>(18, null, n6);
+        n8 = new BinaryTreeNode<>(16, n5, n7);
+
+        // Set up streams
+        System.setOut(new PrintStream(outContent));
```
Comment (brzozan) [inline Assignment2Q1Test.java]: Clever, but it is a bit of an overkill - it is more of an integration test, that checks whole system end to end. Typically though one would not test code by capturing stdout/stderr. Especially stderr will be broken by any debugging messages that one might add to the code in the future.

Integration tests are important, but they are usually used for larger systems where multiple parts cooperate in non-trivial ways. They also give only very high level information about failure that then needs to be tracked down to particular piece of code.
To get more fine-grained information it is better to start with unit tests, testing separate classes directly with assertions like: assertEquals(n8.findAncestors(18), Arrays.asList(...)). Testing in such isolation immediately points to a particular piece of code that might be broken.
Usual practice is to have one test class for each non-trivial implementation class.

Could you please refactor test code to verify directly implementation rather than output?

**Review by brzozan — CHANGES_REQUESTED** [review]
Very good code with extensive test coverage. Good job!

I have only one comment regarding method used for testing. Please see comment in code for details.

**Comment by elifguler96** [comment]
Sorry for the delay, tests should be fine now, I hope 😄

**Comment by elifguler96** [comment]
@brzozan can you also have a look at the changes I made here

File: BinaryTreeNodeTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        n5 = new BinaryTreeNode<>(9, n3, n4);
+        n6 = new BinaryTreeNode<>(19, null, null);
+        n7 = new BinaryTreeNode<>(18, null, n6);
+        n8 = new BinaryTreeNode<>(16, n5, n7);
+    }
+
+    @Test
+    void findAncestorsTest() {
+        // Normal case
+        assertEquals(n8.findAncestors(18), Collections.singletonList(16));
```
Comment (brzozan) [inline BinaryTreeNodeTest.java]: If you put a lot of assertions in a single test method you will end up with errors equivalent to "something's wrong with findAncestors", after which you will have to go through the message in more details to find which one failed.  Plus if the first assertion fails, the other ones will not be executed at all, so you will have to rerun them after fixing first one to find out if something else fails.

Each assertion should be in a separate test method with name describing the condition checked. This will give you trivially understandable error messages like "test such and such failed", immediately telling you which case went wrong. Also with "one assertion per method" approach all tests will be executed and you will get status of all test cases even if some of them fail.

Could you please split the tests to something like `findAncestors_normalCase`, `findAncestors_leaf` etc?

Comment (brzozan) [reply BinaryTreeNodeTest.java]: Also: no need for singletonList. Just Arrays.asList will be sufficient.

File: BinaryTreeNodeTest.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+        // Not in tree case
+        assertEquals(n8.findAncestors(2555), null);
+
+        // Null case
+        assertEquals(n8.findAncestors(null), null);
+    }
+
+    @Test
+    void findCommonAncestorTest() {
```
Comment (brzozan) [inline BinaryTreeNodeTest.java]: Like above, could you please split it to separate one-assertion methods, each annotated with `@Test`?

**Review by brzozan — CHANGES_REQUESTED** [review]
Thank you! Almost there, just one comment to tests :)

File: BinaryTreeNodeTest.java (hunk 11 above)
Comment (elifguler96) [reply BinaryTreeNodeTest.java]: Actually, the IDE suggested `singletonList` but why shouldn't I use it?

**Review by brzozan — APPROVED** [review]
LGTM. Thank you!

File: BinaryTreeNodeTest.java (hunk 11 above)
Comment (brzozan) [reply BinaryTreeNodeTest.java]: After giving it a second thought you are right. `singletonList` is OK here.
It is just quite rarely used method (I don't think I have ever used it or seen it in production code) since it is very, very specialized (immutable list of one element rather than something more general), so it felt a bit odd.

Google code has a strong preference towards things like [ImmutableList.of(...)](http://google.github.io/guava/releases/22.0/api/docs/com/google/common/collect/ImmutableList.html#of-E-) from Guava, which is more general (general purpose immutable list, just created via convenience method taking a single argument, rather than special method to create single-element list); typically other variants of ImmutableList.of(...) would be used to create other, multi-element lists. But that would add another dependency to the project just to provide one class.
